common/credentials/src/coconut/bandwidth/issuance.rs

```diff
@@ -79,141 +53,141 @@ impl BandwidthCredentialIssuanceDataVariant {
-        let type_prehashed = hash_to_scalar(variant_data.info().to_string());
-
-        let params = bandwidth_credential_params();
-        let serial_number = params.random_scalar();
-        let binding_number = params.random_scalar();
+        let params = bandwidth_credential_params().grp();
+        let ecash_keypair = if let Some(id) = identifier {
+            generate_keypair_user_from_seed(params, id)
+        } else {
+            generate_keypair_user(params)
+        };
 
         IssuanceBandwidthCredential {
-            serial_number,
-            binding_number,
             variant_data,
-            type_prehashed,
+            ecash_keypair,
+            expiration_date,
         }
     }
 
     pub fn new_voucher(
-        value: impl Into<Coin>,
         deposit_tx_hash: Hash,
+        identifier: &[u8],
         signing_key: identity::PrivateKey,
         unused_ed25519: encryption::PrivateKey,
     ) -> Self {
-        Self::new(BandwidthVoucherIssuanceData::new(
-            value,
-            deposit_tx_hash,
-            signing_key,
-            unused_ed25519,
-        ))
-    }
-
-    pub fn new_freepass(expiry_date: Option<OffsetDateTime>) -> Self {
-        Self::new(FreePassIssuanceData::new(expiry_date))
+        Self::new(
+            BandwidthVoucherIssuanceData::new(deposit_tx_hash, signing_key, unused_ed25519),
+            Some(identifier),
+            cred_exp_date_timestamp(),
+        )
     }
 
-    pub fn blind_serial_number(&self) -> BlindedSerialNumber {
-        (bandwidth_credential_params().gen2() * self.serial_number).into()
+    pub fn new_freepass(timestamp: u64) -> Self {
+        let exp_timestamp = if timestamp > freepass_exp_date_timestamp() {
+            error!(
+                "the provided free pass request has too long expiry, setting it to max possible"
+            );
+            freepass_exp_date_timestamp()
+        } else {
+            timestamp
+        };
+        Self::new(
+            BandwidthCredentialIssuanceDataVariant::FreePass,
+            None,
+            exp_timestamp,
+        )
     }
 
-    pub fn blinded_serial_number_bs58(&self) -> String {
+    pub fn ecash_pubkey_bs58(&self) -> String {
         use nym_credentials_interface::Base58;
 
-        self.blind_serial_number().to_bs58()
+        self.ecash_keypair.public_key().to_bs58()
     }
 
     pub fn typ(&self) -> CredentialType {
         self.variant_data.info()
     }
 
-    pub fn get_private_attributes(&self) -> Vec<&PrivateAttribute> {
-        vec![&self.serial_number, &self.binding_number]
-    }
-
-    pub fn get_public_attributes(&self) -> Vec<&PublicAttribute> {
-        vec![self.variant_data.public_value(), &self.type_prehashed]
-    }
-
-    pub fn get_plain_public_attributes(&self) -> Vec<String> {
-        vec![
-            self.variant_data.public_value_plain(),
-            self.typ().to_string(),
-        ]
+    pub fn expiration_date(&self) -> u64 {
```

Comment:
ibid.

Reply:
See earlier comments on timestamp handling.
TBD

Reply:
TO FIX : timestamps

Reply:
Fixed in 7bc7a8dde1e5c49778ba132e454a138dd1f0fdea